src/wasm/wasm.cpp

```diff
@@ -362,41 +301,55 @@ static void handleUnreachable(Block* block,
 }
 
 void Block::finalize() {
+  if (list.size() == 0) {
+    type = Type::none;
+    return;
+  }
   if (!name.is()) {
-    if (list.size() > 0) {
-      // nothing branches here, so this is easy
-      // normally the type is the type of the final child
-      type = list.back()->type;
-      // and even if we have an unreachable child somewhere,
-      // we still mark ourselves as having that type,
-      // (block (result i32)
-      //  (return)
-      //  (i32.const 10)
-      // )
-      if (type.isConcrete()) {
-        return;
-      }
-      // if we are unreachable, we are done
-      if (type == Type::unreachable) {
+    // nothing branches here, so this is easy
+    // normally the type is the type of the final child
+    type = list.back()->type;
+    // and even if we have an unreachable child somewhere,
+    // we still mark ourselves as having that type,
+    // (block (result i32)
+    //  (return)
+    //  (i32.const 10)
+    // )
+    if (type.isConcrete()) {
+      return;
+    }
+    // if we are unreachable, we are done
+    if (type == Type::unreachable) {
+      return;
+    }
+    // we may still be unreachable if we have an unreachable
+    // child
```

Comment:
It might be worth pointing out explicitly that only blocks that would otherwise have `Type::none` are eligible to become `unreachable` if they have an unreachable child. The previous two `if`s could even be replaced with `if (type != Type::none)`. Also, some of this work looks duplicated with work done in `handleUnreachable`. Can that function be used here?

Reply:
Good point! This is pre-existing code that was just de-indented, so I didn't read it carefully til now. Yes, it seems like it does duplicate the other pre-existing code... will fix.